sphinx/search/__init__.py

```diff
@@ -168,8 +168,17 @@ class _JavaScriptIndex:
     SUFFIX = ')'
 
     def dumps(self, data: Any) -> str:
-        data_json = json.dumps(data, separators=(',', ':'), sort_keys=True)
-        return self.PREFIX + data_json + self.SUFFIX
+        def json_dump(data: str | tuple | list) -> str:
+            return json.dumps(data, separators=(',', ':'))
+
+        js_indices = {
+            json_dump(name): f'new Map({json_dump(sorted(entries.items()))})'
+            if isinstance(entries, dict)
+            else json_dump(entries)
+            for name, entries in sorted(data.items())
+        }
+        data_js = '{' + ','.join(f'{k}:{v}' for k, v in js_indices.items()) + '}'
+        return self.PREFIX + data_js + self.SUFFIX
```

Comment:
Could we use a [jinja](https://jinja.palletsprojects.com/en/stable/) template for the search index? That might make this look a little better. Seems to be one of Sphinx's dependencies: https://github.com/sphinx-doc/sphinx/blob/80ed1b4cd1a5290106f2bff528b27f0bf16fb32a/pyproject.toml#L65

Definitely worth adding a comment or two explaining what's going on.

Reply:
I'm mulling over the suggestion; making the logic and output format of the method clearer would both be helpful, but I'd probably prefer not to include an additional technology in the mix (even though as you mention, it's already available as an application dependency).

Reply:
I think we should use it if it's available and it makes things clearer. It's usage elsewhere in Sphinx is pretty extensive: https://github.com/search?q=repo%3Asphinx-doc%2Fsphinx%20jinja&type=code

Reply:
(although dumping the index as JSON is perhaps a better alternative)

Reply:
@wlach I'm fairly sure Jinja was originally written for use in Sphinx / Flask -- they're all originally Pocco projects!